proxy/router/src/cache.rs

```diff
@@ -1,88 +1,239 @@
 use indexmap::IndexMap;
-use std::hash::Hash;
+use std::{hash::Hash, ops::{Deref, DerefMut}, time::{Duration, Instant}};
 
 // Reexported so IndexMap isn't exposed.
 pub use indexmap::Equivalent;
 
-/// A cache for routes
+/// An LRU cache
 ///
 /// ## Assumptions
 ///
 /// - `access` is common;
 /// - `store` is less common;
-/// - `capacity` is large enough..
+/// - `capacity` is large enough that idle vals need not be removed frequently.
 ///
 /// ## Complexity
 ///
 /// - `access` computes in O(1) time (amortized average).
 /// - `store` computes in O(1) time (average).
-// TODO LRU
-pub struct Cache<K: Hash + Eq, V> {
-    vals: IndexMap<K, V>,
+/// - `reserve` computes in O(n) time (average) when capacity is not available,
+pub struct Cache<K: Hash + Eq, V, N: Now = ()> {
+    vals: IndexMap<K, Node<V>>,
     capacity: usize,
+    max_idle_age: Duration,
+
+    /// The time source.
+    now: N,
+}
+
+/// Provides the current time within the module. Useful for testing.
+pub trait Now {
```

Comment:
When we're using the new `tokio`, this probably could be `tokio_timer::Now`, or at least implemented for `tokio_timer::Now`...